### DISCUSSION TITLE
Syncing URL with algolia instantsearch using next.js adding one unnecessary re-route

### DISCUSSION THREAD

**Discussion by AbdelrahmanAhmed605** [vercel/next.js#61593] [Help]
### Summary

I am not really sure if this is a next.js issue or an algolia instantsearch issue but I thought I might give this a try on this repo. 

I am using next.js with algolia instantsearch and currently I have a regular form input on my homepage that allows a user to enter a search. I made it so that when a user submits their search on the homepage, it uses next.js router to re-direct them to the /search page which uses algolia's instantsearch with the search query already applied (to do this, I used routing={true} on my InstantSearch component). 

The issue I have comes from the browser history. So when a user types "phone" on the home-page, it goes from "/" -> "/search?search=phone" -> "/search?<index_name>%5Bquery%5D=phone". Now when the user clicks the back button on the browser, it takes them to "/search?search=phone" which isn't actually connected to algolia, so it just shows the /search page with all the results in my index. How can I have it so that the user is instantly taken to "/search?<index_name>%5Bquery%5D=phone" instead of "/search?search=phone" so that when they press back on the browser, it takes them to the homepage "/"?

### Additional information

```js
### For Context

On my homepage, I have the following event listener

import { useRouter } from 'next/navigation';

const router = useRouter();
const handleSearch = (e) => {
    e.preventDefault();
    router.push(`/search?search=${encodeURIComponent(query)}`)
  };
```

On my search page.tsx, I have:
```
import { useSearchParams } from 'next/navigation';

const SearchPage = () => {
  const searchParams = useSearchParams();
  const search = searchParams.get('search');

  return (
    <InstantSearch
      searchClient={searchClient}
      indexName={indexName}
      initialUiState={{
        [indexName]: {
          query: search,
        },
      }}
      future={{
        preserveSharedStateOnUnmount: true,
      }}
      routing={true}
    >
      <SearchComponent />
    </InstantSearch>
  );
};
```
```


### Example

_No response_

**Comment by ega123** [comment]
Have you found an answer for this issue yet?

**Reply by pkumarRB** [reply]
I have observed a similar issue where it takes two clicks on the browser’s Back button to navigate back to the home page.

**Comment by paramjeetn** [comment]
To eliminate the intermediate history step (`/search?search=phone` -> `/search?<index_name>[query]=phone`), perform the query mapping **before** pushing to the router from your homepage form:

### Solution:

1. **Construct the Algolia-compatible query param directly on form submission:**
```tsx
'use client'

import { useRouter } from 'next/navigation'
import { useState } from 'react'

export function SearchForm() {
  const [query, setQuery] = useState('')
  const router = useRouter()

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault()
    if (!query.trim()) return

    // Encode directly into the key format InstantSearch expects:
    const params = new URLSearchParams()
    params.set('your_index_name[query]', query)

    // Navigate directly to the final URL:
    router.push(`/search?${params.toString()}`)
  }

  return (
    <form onSubmit={handleSubmit}>
      <input 
        value={query} 
        onChange={(e) => setQuery(e.target.value)} 
        placeholder="Search..." 
      />
    </form>
  )
}
```

2. **Ensure `InstantSearch` uses `router.replace` internally during refinement:**
In your search page configuration, use `history.replaceState` or Next.js router `replace` in the custom router mapping so subsequent in-page filter refinements do not clutter the back-button stack.
